### ISSUE TITLE
Apply pinned-host staging + async D2H to MemmapBackend.write_emit (close the residual↔emit asymmetry)

### ISSUE THREAD

**Issue by AlliedToasters** [AlliedToasters/fpwap#63]
## Summary

`seg.buffer.write_slice` (residual path) clocks ~1ms per microbatch on a 70B run because it uses pinned-host staging + async GPU→host copy from #44. `storage.write_emit` (emit path) clocks ~50ms per microbatch on the same data shape because it does a synchronous `tensor.detach().cpu()` followed by a synchronous host→memmap memcpy. Same data, same disk, same memmap — the only difference is whether the GPU→host transfer is on the critical path.

Filed as the follow-up to #61 / #62. #62 removed the per-write *amplifier* (msync + fadvise) that made emit climb 76→464ms. What's left is the ~50ms floor described here, which is structural: the emit write path bypasses the staging machinery that already exists for the residual path.

This is the architectural step that closes the gap.

## Diagnostic data (today, llama-3.3-70B, RTX 5090 / 128 GB host, post-#62)

Config: `detect_layers=[2]`, lazy forward (#48) exits after fpwap layer 1, layer 1 ends up last-in-chunk so #59's per-microbatch postfix fires. 10660 samples × seq_len 1025 × hidden 8192 × bf16 → 2668 microbatches × 67 MB residual + 67 MB emit each.

Stable per-microbatch timings observed across the layer (deep into the run, page cache pressured):

| Metric | Per microbatch | Path |
|---|---|---|
| `write` | 1-2 ms | `seg.buffer.write_slice` → pinned-host staging + async D2H (#44) + memmap write returns when *enqueued* |
| `emit` | 40-80 ms | `storage.write_emit` → `_Shard.write` → `tensor.detach().cpu()` (sync, blocks on transfer) + host→memmap memcpy |

`emit` is no longer climbing (that was the per-write `_flush_and_evict()` removed in #62), but it sits ~40× above `write` because the GPU→host transfer is on the critical path.

PCIe Gen4 x16 sustained ~16-20 GB/s practical, shared with weight streaming. 67 MB / ~10 GB/s under contention ≈ 7-30ms transfer + ~10ms host→memmap memcpy = ~50ms — consistent with the observed floor.

## Where the asymmetry lives

`fpwap/storage/memmap.py:86-92` (`_Shard.write`):

```python
def write(self, sample_ids: Tensor, tensor: Tensor) -> None:
    mm = self._ensure(tensor)
    ids_np = sample_ids.cpu().numpy()
    host = tensor.detach().cpu()           # SYNC GPU→host
    if self._stores_bf16_as_u16:
        host = host.view(torch.uint16)
    mm[ids_np] = host.numpy()              # SYNC host→memmap memcpy
    self._dirty = True
```

vs the residual path (in `ResidualBuffer.write_slice` from #44 / #43): the same data is queued through a pinned-host staging tensor with `non_blocking=True`, and the actual host→memmap drain runs concurrently with the next microbatch's compute via the CUDA copy engine.

`storage.write_emit` uses none of that. It also doesn't allocate a pinned host buffer, so even the GPU→host copy that does happen can't use the high-throughput pinned path.

## Proposed approach

Mirror the residual buffer's design inside `MemmapBackend`:

1. **Pinned-host staging tensor per `_Shard`** — bounded size (e.g. configurable, default ~2-8 GB). Allocated lazily on first write.
2. **Non-blocking GPU→pinned-host copy** — `staging[start:stop].copy_(tensor, non_blocking=True)` returns immediately.
3. **Background worker drains pinned-host → memmap** — already-merged emit-drain-at-chunk-boundary semantics from #62 fit here cleanly; the chunk-boundary `drain_emits()` waits on the staging-buffer copy stream and then issues the (now CPU-resident) memmap writes.
4. **Backpressure when staging is full** — if the staging tensor would overflow before the next chunk drain, block the next `write` until the worker has drained enough, same as residual.

Existing #62 plumbing — `_dirty` flag, `drain_emits()`, `_flush_and_evict` deferred to chunk boundary — stays as-is. This change replaces the body of `_Shard.write` with the staging-enqueue path, and `drain_emits()` gains a step that flushes the pinned-host staging into the memmap before the existing msync.

## Expected impact

- `emit` per microbatch drops from ~50ms to ~1-2ms (match `write`).
- For our 70B 10k-sample run with one captured layer: ~133s saved per captured layer (50ms × 2668 microbatches).
- For the paper-typical 80-layer 70B with one middle-layer probe: same per-layer savings, applied once per captured layer per dataset. On large multi-eval / multi-control runs (the dd-repo concat-everything pattern), this compounds.
- Crucially, this path becomes architecturally *symmetric* with residual — no \"capture is the slow special case.\" Speed-to-reproduction parity for users who care about extracted activations vs. just running forward passes.

## Why this matters for downstream consumers

The dd-repo fork on this branch (`feat/fpwap-integration`) is built around the premise \"speed-to-reproduction is the only metric that matters.\" The whole point of running fpwap over HF on a single consumer GPU is to reproduce the deception-detection paper's probes as fast as possible. With #62 landed, the residual path is fully optimized; emit is now the long pole on captured-layer runs. Closing this gap would put fpwap's captured-layer wall time within ~5% of an uncaptured forward pass — which is the bar that makes this regime really sing.

## Out of scope

- Multi-layer capture optimizations (different problem; separate issue if/when it bites).
- Tuning the staging buffer size — defaults should be fine for the 70B regime, may need a knob later for memory-constrained hosts.
- Using GPU-resident scratch for emits (the residual path uses scratch within a chunk; emits could in principle do the same, but pinned-host staging is the cheaper structural change and matches #44's pattern).

## Repro / validation

Same shape as #59 / #61:

- Model: llama-70B-class
- Dataset size: 10k+ samples × seq 1025 × hidden 8192 × bf16
- `buffer_path=` set, `offload_dir=` set
- Capture at any single layer (use lazy forward + last-in-chunk for clean per-microbatch postfix)
- Before patch: `emit` postfix shows ~50ms per microbatch, flat
- After patch: `emit` postfix shows ~1-2ms per microbatch, flat — matching `write`
- `Profile.emit_s` should drop by ~25× on a captured layer
